src/util/obbt.jl

```diff
@@ -0,0 +1,406 @@
+        error(LOGGER, "the option upper_bound_constraint cannot be set to true without specifying an upper bound")
+    end 
+
+    if !isinf(rel_gap) && isinf(ub)
+        error(LOGGER, "rel_gap_tol is specified without providing an upper bound")
+    end
+end 
+
+function constraint_obj_bound_polynomial(pm::GenericPowerModel, bound)
+    PowerModels.check_polynomial_cost_models(pm)
+
+    from_idx = Dict(arc[1] => arc for arc in ref(pm, :arcs_from_dc))
+
+    @constraint(pm.model, 
+            sum(   
+                gen["cost"][1]*var(pm, :pg, i)^2 +
+                gen["cost"][2]*var(pm, :pg, i) +
+                gen["cost"][3]
+            for (i,gen) in ref(pm, :gen)) +
+            sum(   
+                dcline["cost"][1]*var(pm,:p_dc, from_idx[i])^2 +
+                dcline["cost"][2]*var(pm, :p_dc, from_idx[i])  +
+                dcline["cost"][3] 
+            for (i,dcline) in ref(pm, :dcline))
+            <= bound 
+    )
+end 
+
+function constraint_obj_bound(pm::GenericPowerModel, bound)
+    model = PowerModels.check_cost_models(pm)
+    if model != 2 
+        error("Only cost models of type 2 is supported at this time, given cost model type $(model)")
+    end 
+    constraint_obj_bound_polynomial(pm, bound)
+end 
+
+function create_modifications(pm::GenericPowerModel, 
+    vm_lb::Dict{Any,Float64}, vm_ub::Dict{Any,Float64}, 
+    td_lb::Dict{Any,Float64}, td_ub::Dict{Any,Float64}) 
+    
+    modifications = Dict{String,Any}()
+
+    modifications["per_unit"] = true
+    modifications["bus"] = Dict{String,Any}()
+    modifications["branch"] = Dict{String,Any}()
+
+    buses = ids(pm, :bus)
+    buspairs = ids(pm, :buspairs)
+    branches = ids(pm, :branch)
+
+    for bus in buses
+        index = string(ref(pm, :bus, bus, "index"))
+        modifications["bus"][index] = Dict{String,Any}( "vmin" => vm_lb[bus], "vmax" => vm_ub[bus] )
+    end 
+
+    for branch in branches 
+        index = string(ref(pm, :branch, branch, "index"))
+        f_bus = ref(pm, :branch, branch, "f_bus") 
+        t_bus = ref(pm, :branch, branch, "t_bus")
+        bp = (f_bus, t_bus)
+        modifications["branch"][index] = Dict{String,Any}( "angmin" => td_lb[bp], "angmax" => td_ub[bp] )
+    end 
+
+    return modifications
+end 
+
+function create_modifications(pm::GenericPowerModel, lb::Float64, ub::Float64, id::Int64)
+    
+    modifications = Dict{String,Any}()
+
+    modifications["per_unit"] = true
+    modifications["bus"] = Dict{String,Any}()
+
+    index = string(ref(pm, :bus, id, "index"))
+    modifications["bus"][index] = Dict{String,Any}( "vmin" => lb, "vmax" => ub )
+
+    return modifications
+
+end 
+
+function create_modifications(pm::GenericPowerModel, lb::Float64, ub::Float64, id::Tuple{Int64,Int64})
+
+    modifications = Dict{String,Any}()
+
+    modifications["per_unit"] = true
+    modifications["branch"] = Dict{String,Any}()
+
+    branches = ids(pm, :branch)
+
+
+    for branch in branches 
```

Comment:
Putting this loop in hear will result in a `|branches|^2` runtime to update all of the branches.  Given the style of create_modifications for buses, I would recommend that the `id` in this function be the branch id.  The 1-to-many mapping for pairs to branches should be handled in the loop that calls this function.

Reply:
removed in lieu of sequential_obbt removal